### DISCUSSION TITLE
Reactors are broken (and possible rework inside)

### DISCUSSION THREAD

**Discussion by ubertpendragon** [FakeFishGames/Barotrauma#8702] [Unstable feedback]
Since there's a power rework in the works and Engineer gameplay is pretty limited, it's time to take a look at Reactors:

1. **Auto-reactor fails to output the highest amount of power**
and
2. **Manual visual indicators lie**

On the first image, we have a Reactor with 10000kw max output. 
The load is 10000 and the auto-reactor tries its best to get as much power out as possible, pushing the needle in the green. It produces 5600kw.
On the second image, we have the same reactor producing 8500kw.

![image](https://user-images.githubusercontent.com/73229309/161387609-06e965dd-bb53-42b6-b1aa-9685080477f1.png)
![image](https://user-images.githubusercontent.com/73229309/161387613-4df25ad6-a880-4c85-af24-4a866ee758f5.png)

Not only the auto-reactor fails to meet its maximum output, the optimal Turbine rate is not what the reactor indicates (presumably green), 85% is. Using one uranium rod, below and above 85%, you lose output, which is counter-intuitive.

---

3. **Fuel rod "power" scales with reactor Max Output**

As we've seen, a single Uranium Fuel Rod can generate 8500kw on a 10000kw max output reactor.
On a 20000kw max output reactor, the same rod will output 17000kw

Thorium and Fulgurium rod allow you to push turbine to 100 and **a single rod will _always_ produce more power than the Max Output cap!**. A single of those rods goes through the "roof" of max power output. Reactors thus have 4 rod slots that only exist for dividing the durability between rods.

Increasing max output through upgrades (+10%) will also (slightly) increase the fuel efficiency of a reactor, as it uses less heat for the same amount of power.

You normally would expect bigger submarines and larger reactors to require more fuel rods to function, but they don't.
A submarine that draws 5000kw of load on average with a 10000kw max output reactor will consume the same amount of fuel as a 10000kw average load behemoth with a 20000kw max output reactor.

Max reactor output should only determine how much power can be distributed across the sub; not how powerful fuel rods are, nor how long they last.
The only fine-tuning variable is "Fuel Efficiency" which dictates how fast a fuel rod is consumed. There is no variable for not making fuel rods twice as powerful on larger reactors.

---

4. **Turbine knob is useless**

Keeping this previous turbine rate in mind, you can slap a MEM with 85 directly into the _Set_TurbineOutput_ of your reactor and *voila*.
Only fission rate matters. Your reactor only needs one knob to have control the power output and prevent fires. 
In appearance, this makes reactor control harder than it actually looks. 

Using one uranium rod, you can leave turbine at 85% for "optimal" power generation, but you'll still be capped below max output.
Using any other amount/type of rods, slap turbine to 100% and never touch it again. 
You only need to pick either 85 or 100 and ignore the knob. 

So you can technically only use fission, but the knob is also really finnicky, lacking precision.

---

5. **Everybody hates Manual**

Manual is, simply put, not fun. The only rare, practicale use of manual is restarting a reactor. 
There is no reason to switch to manual outside of that, juggling with 2 knobs is difficult (well, 1 really) and it's something you do not want to do for more than 10 seconds.

---

6. ARCs are overpowered

Automated Reactor Controllers or ARC basically remove overvolting, undervolting and JB fires as game mechanic on top of making your sub more fuel-efficient for the cost of 2 components. 
They are essentially Auto on steroids, and are far more precise and responsive than manual can ever be. 
You can see how this is not balanced.

---

# **SUGGESTED REWORK**

![ReactorRework](https://user-images.githubusercontent.com/73229309/161395100-fc33e2f3-7d3b-4b1e-999c-ec5a15b88ff8.jpg)

**Goals:**
- Fuel rod output and efficiency stops scaling with max reactor output
- Manual control is easier to use
- Benefits of manual control (critical output), more engineer gameplay
- More mixing and matching of fuel rods

## **Other ideas**:

Overvolting changes to make the mechanic less frustrating/punishing. One of the reasons why players use ARCs.
https://github.com/Regalis11/Barotrauma/discussions/6870

Reasons to overvolt/undervolt: more engineer gameplay.
Overvolting to "burn" ballast flora (ballast flora takes damage from overvolted JB)
Undervolting to resist EMPs (undervolted items do not take EMP damage)

**Comment by LePiromano** [comment]
i agree

**Comment by Regalis11** [comment]
Thank you for the feedback! While I agree with pretty much everything you wrote, I still don't agree with your first point.

> Auto-reactor fails to output the highest amount of power
> Manual visual indicators lie

In my view, trying to push the reactor to it's theoretical maximum is a sort of edge case that the automatic control doesn't need to, or maybe even shouldn't, be able to handle. I also wouldn't say the optimal range indicator is lying: it shows the theoretical optimal value (the optimal value in this case being "as high as possible"), but in practice push the turbine output to it's very limit causes the reactor to cool down unless you take measures to prevent that. As you can see in the screenshot, the fission rate can't go any higher, and you need to address that by adding more fuel.

**Reply by ubertpendragon** [reply]
>  shouldn't, be able to handle

The 1 Uranium rod + Auto configuration is how most pub subs are powered; and not getting the maximum amount of power out of that feels like kind of a _noob-trap_.

Ultimately, the things the Auto-reactor "shouldn't be able to handle" are power/load spikes but I still don't see anyone going manual for it unless it's a very predictable function of the sub.
If there were more reasons to go manual maybe it'd be worth considering, overheating the reactor to get extra juice was my take on it, as well as deliberate Under/Overvolting

**Reply by cshingle** [reply]
This may be what @Regalis11 is getting at, but I think a reasonable way to add game play would be to make the Auto-reactor do a reasonable job with power/load spikes, but not be able to driver the reactor as hard as manual mode to add some incentive for players to "go manual" during a crisis.
The issue I see with that that people will continue to just wire the reactor to obtain max output. A way to deal with this is to allow manual mode to drive the fission rate past %100 (safe output) in exchange for more damage, fire risk and maybe a efficiency penalty (burn rods faster). You could allow wiring to push it past %100 (safe) output, but it is not something that would be a set it and forget it like now.

On a slightly related note it would be nice if there where additional "passive" creatures that would attach itself to the hull like a Remora Fish but instead of catching a ride it sap some power from nearby equipment. The idea of this creature is that it would require someone to keep an eye on the power system even on subs that have reactor headroom.

**Reply by MateusJK15** [reply]
Another good way to incentive players to use the manual mode is to make it more complex, more challenging, like adding more gauges, more settings to mess with (like a real life reactor), because as it is now it's just 2 sliders that go left and right, there's no challenge and absolutely no fun.

**Comment by mLuby** [comment]
I agree with the problems @ubertpendragon raises. An alternate solution:

> Limit automatic control¹ of the reactor to ±20%²³ of its last manual setting and require 10 seconds³ of stability⁴ at the new manual setting before automatic reactor controls can take over again.

Thus significant power demand changes benefit from _brief_ _manual_ control of the reactor, encouraging communication between navigation and engineering like "all ahead full!" For example, going from engines stopped to engines 100% will undervolt until an engineer adjusts the reactor—the sub will still function without that manual intervention, just not optimally. (As noted, overvolting is much more punishing but luckily many subs already include a Shutdown/SCRAM button on the Navigation Terminal to help with that situation; regardless captains will adapt quickly). As for the auto GUI switch vs an ARC circuit, the latter's improved responsiveness over the former means it's still a worthy in-game project for engineers but doesn't entirely solve reactor gameplay like both GUI and circuit do now.

¹ Meaning _both_ the auto GUI switch and the set_fission/set_turbine inputs that ARC circuits use.
² Ideally show this range in the load chart, maybe something like the edited image below.
³ These settings should be exposed to sub builders, which has the added benefit of letting groups "disable" this feature per sub by setting ±100%/0s.
⁴ An engineer might need to manually adjust the reactor for a few seconds until auto takes over. One way to implement this could counting from 10s to 0s as long as fission/turbine rounded to ten remains the same, then enabling ARCs—if it does change in that time, reset the counter.
![reactor](https://user-images.githubusercontent.com/2483420/161798872-20e297fb-8dd7-4da3-8699-2e7308325843.png)

**Comment by Oukido** [comment]
Firstly I'd like to mention that I am not a fan of punishing good designed reactor controllers, since I really enjoyed my time building those on my own. It gave me a sense of achievement when I first built one that did well. That said I do not mean the 2 component regex one which rapes the reactor... it is an abomination in my mind.

Additionally keep in mind you basically need to already be a pro in this game to even think about fiddling with the reactor manually, since if you are not you have lots to do with straight up surviving. (of course not everyone is the same, some struggle more than others)

I suggest instead of nerfing the autocontrollers give actual alternatives with manual control. Like others already mentioned the controls are straight up bad and very unintuitive and probably a large portion of the playerbase does not even know you could control it with WASD, which to be fair does not really make it more usable.
Instead why not rebuild the UI with vertical gauges similar to the heat gauge and a vertical slider besides it which visually correlates with the target-value you want to set. This would help alot with the intuitiveness.

Now what about the alternative to auto controllers. The really big advantage of auto control like what a PID provides is that it can overshoot the target to speed up the needle faster towards the target and then snap back. And it can also do that in situations where power demand and power supply are heavily fluctuating. A player has a really hard time to compete with that just by adjusting the sliders.

So why not add impactful, Player/Bot only usable buttons to quickly push the needles in one or the other direction. For example:

- A button to vent some steam out of the reactor to quickly drop Heat and power generation. - in case of load drop
- A button to inject incendium into the reactor chamber to quickly rise heat and fission rate. - in case of low heat.
- A button to decouple the turbine generator to increase rpm speed (no magnetic friction). - in case of load rise (uses up some internal power buffer that needs to be recharged)

If it is desired there could be added slots for actual items to be consumed like incendium and maybe batteries for charge (or even alien power cells?)

So if someone decides to not use auto controllers in their gameplay they can actually do it effectively.

**Comment by Renko-6662** [comment]
Very good idea as always. I'm agreeing with basically everything.

I feel it just lack one things : radiations.

For example, it could work like that : Every time the reactor is damaged, it should lose some of its max durability. When max durability is below a set threshold, the reactor would start to generate radiations randomly. You could repair the max durability by having to craft some special item (kinda like the repair pack, but I think not talent-gated).

Or we could have the reactor separated into smaller pieces : the turbine, the rod containment chamber, etc.
In this case, only the rod containment chamber would generate radiation when damaged, for example.

**Comment by doornydew** [comment]
as an engineer main the things i specifically am looking for would be:

getting skills whilst working on/with the reactor, or even its own skill that could possibly interact some way to the reactor (make it so theres truly a reactor engineer, and an electrical one)
im talking more out of each rod, eventually getting like a blue? (or whatever color) section in the gauges to show optimal heat or power generation (bare minimum + - a tiny bit to make it easier to waste as little juice as possible)

alot more intuitive UI or atleast something more to do than look at 2 sliders... make me feel like im being actually usefull.

radiation would be a neat thing, i can understand people not wanting it much but...  its a reactor, and theres like no way to get irradiated unless you  either fuck up, or blow yourself up intentionally.

maybe have a limiting factor on how quickly outside automation could adjust the reactor, as in, manual will always be faster, but you could then still use automation to, for instance, undervolt and underheat at more "calm" moments, making it worthwhile to GET automation, but it can never replace a human (also buffs reactor PDA quite nicely)
